### DISCUSSION TITLE
with_test_writer() doesn't work in  unit tests

### DISCUSSION THREAD

**Discussion by realtica** [tokio-rs/tracing#2736] [Q&A]
Hello, I do not know why Tracing only fails inside unit tests, here is my config:

```
tracing = "0.1.37"
time = "0.3.29"
tracing-subscriber = {version = "0.3.17", features = ["env-filter", "fmt", "ansi", "std", "time"] }
```

```
    let time_format = time::format_description::parse("[hour]:[minute]:[second]")
        .expect("format string should be valid!");
    let timer = UtcTime::new(time_format);
    let format = fmt::format()
        .with_line_number(true)
        // .with_source_location(true)
        .with_timer(timer)
        .with_level(true) // don't include levels in formatted output
        .with_target(false) // don't include targets
        .with_thread_ids(false) // include the thread ID of the current thread
        .with_thread_names(false) // include the name of the current thread
        .pretty(); // use the `Compact` formatting style.
    tracing_subscriber::fmt()
        .with_test_writer()
        .event_format(format)
        .init();

```

**Comment by hawkw** [comment]
You are calling the `fmt::Subscriber::init()` method, [which panics if it is called more than once, as discussed in the documentation](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/util/trait.SubscriberInitExt.html#method.init). Since multiple unit tests will run in the same binary, each test that calls into your `tracing` setup code will call `init()`, and any subsequent calls to `init()` after the first will panic. You can fix this by using [`try_init()`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/util/trait.SubscriberInitExt.html#method.try_init) rather than `init()` --- this method returns a `Result` if the global default subscriber has already been initialized. In tests, you can simply ignore the `Result` this method returns.

Hope that helps!

**Reply by realtica** [reply]
Not showing any log in unit tests yet, I also added tracing-log feature too.


```
let _ = tracing_subscriber::fmt()
        .with_test_writer()
        .event_format(format)
        .try_init();
```
`tracing-subscriber = {version = "0.3.17", features = ["env-filter", "fmt", "ansi", "std", "time", "tracing-log"] }`

```
❯ cargo test
   Compiling rapidsite v0.0.1 (/home/realtica/projects/rapidsite/src-tauri)
    Finished test [unoptimized + debuginfo] target(s) in 6.20s
     Running unittests src/main.rs (target/debug/deps/rapidsite-91a95a5b19e1c261)

running 2 tests
test app::file_management::tests::get_folder_list_test ... ok
test app::file_management::tests::get_file_list_test ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

**Reply by hawkw** [reply]
This is normal. By default, Rust tests capture all output from the test and only display it if the test fails. If you want to see output from passing tests, run your tests using 
```shell
cargo test -- --show-output
```
